pine64-pinephonepro: Init

wentam opened the pull request:
Depends on #516 and #522. Based on #445.

Continuing work on pine64-pinephonepro.

This is intended to be used with the guided installer (#522).

## Compared to #445 (at time of writing)

Behind:
* Commits unrelated to PPP removed.

Ahead:
* kernel 5.16.7 -> 5.17.9
* kernel patch to fix pipewire/wireplumber segfault in libcamera
* kernel patch to wake from suspend on phone call
* kernel patch for pine64 keyboard case
* kernel configuration needed for functional firewall
* now uses #516 for managing UCM profiles
* pipewire gets UCM profiles
* kernel config to get waydroid working with networking (with hardware accel if you use lineageOS 18.1+)

## Tested working

- [x] GPU
- [x] Audio playback/recording (pipewire and pulse both)
- [x] Wifi
- [x] Bluetooth
- [x] Mobile data
- [x] GPS (need to inject assistance data, poor reception likely a result of poor antenna/hardware)
- [x] Suspend
- [x] Making calls
- [x] Recieving calls
- [x] Recieving calls from suspend (usually works. Not 100% reliable, but that's probably a userspace problem.)
- [x] Sending texts
- [x] Recieving texts
- [x] Touch screen
- [x] Notification LED
- [x] Power/volup/voldown buttons
- [ ] Cameras (Awaiting kernel support in cross-distro kernel)
- [x] Accelerometer/Gyroscope mpu6500
- [ ] Geo Magnetic Sensor af8133j (Awaiting kernel support in cross-distro kernel)
- [ ] Ambient light sensor stk3311 (Unknown, kernel driver enabled, I don't have /sys/bus/iio device for it)
- [x] pine64 keyboard accessory
- [x] Charging reliably (though for keyboard case you need to run `echo 1500000 > /sys/class/power_supply/rk818-usb/input_current_limit` https://xnux.eu/pinephone-keyboard/faq.html)
- [x] waydroid
- [x] Flashlight
- [x] Vibrate motor
- [x] firewall 

## TODO
* README
* release.nix build installer?
* mobile.documentation.hydraOutputs? (see devices/pine64-pinephone/default.nix)

## menuconfig note

I needed the following patch to make menuconfig work. This is not commited as I assume it may break other devices:

```
diff --git a/overlay/mobile-nixos/kernel/builder.nix b/overlay/mobile-nixos/kernel/builder.nix
index 4c72aeab..caad9931 100644
--- a/overlay/mobile-nixos/kernel/builder.nix
+++ b/overlay/mobile-nixos/kernel/builder.nix
@@ -539,7 +539,8 @@ stdenv.mkDerivation (inputArgs // {
     menuconfig = kernelDerivation.overrideAttrs({nativeBuildInputs ? [] , ...}: {
       nativeBuildInputs = nativeBuildInputs ++ [
         ncurses
-        pkg-config-helper
+        ncurses.dev
+        pkg-config
       ];
       buildFlags = [ "nconfig" "V=1" ];
 
@@ -547,14 +548,6 @@ stdenv.mkDerivation (inputArgs // {
       buildPhase = ''
         (PS4=" $ "; set -x
 
-        # Hot fixes pkg-config use.
-        export PKG_CONFIG_PATH="${buildPackages.ncurses.dev}/lib/pkg-config"
-        if [ -e scripts/kconfig/nconf-cfg.sh ]; then
-          sed -i"" \
-            -e 's/$(pkg-config --libs $PKG)/-L $(pkg-config --variable=libdir ncursesw) $(pkg-config --libs $PKG)/' \
-            scripts/kconfig/nconf-cfg.sh
-        fi
-
         cat >> scripts/kconfig/Makefile <<EOF
 
         run-nconfig:
```

samueldr committed:
hardware-rockchip: Add rockchip-rk3399s

samueldr committed:
pine64-pinephonepro: Init kernel

Config is the defconfig from mobian, arbitrarily changed.

samueldr committed:
pine64-pinephonepro: Add firmware

This is based on the pinebook pro firmware package. Unknown of this is
all of the required firmware, and if there's too much.

samueldr committed:
pine64-pinephonepro: Add USB role switch task

This is used to force the type-c direction.

samueldr committed:
pine64-pinephonepro: Init

samueldr committed:
pine64-pinephonepro: Enable default LED behaviour

samueldr committed:
pine64-pinephonepro: WIP kernel config validation

samueldr committed:
pine64-pinephonepro: WIP shell on stage-1 fail

wentam committed:
pine64-pinephonepro: add alsa UCM profiles

wentam committed:
pine64-pinephonepro: Update kernel to 5.16.0

And change origin to the now decided location of the community
maintained kernel.

wentam committed:
pine64-pinephonepro: vibrate motor kernel support

samueldr committed:
pine64-pinephonepro: kernel 5.16.0 → 5.16.7

samueldr committed:
pine64-pinephonepro: Use appropriate frequency scaling

See https://gitlab.com/pine64-org/linux/-/merge_requests/36

TLDR: chip too hot, designed to run cooler slower.

samueldr committed:
pine64-pinephonepro: ensure SPI config is enabled

samueldr committed:
pine64-pinephonepro: Normalize kernel config

samueldr committed:
pine64-pinephonepro: Add charging workarounds

See:

 - https://gitlab.com/pine64-org/linux/-/merge_requests/34

wentam committed:
pine64-pinephonepro: runCommandNoCC -> runCommand

wentam committed:
pine64-pinephonepro: kernel 5.16.7 -> 5.17.9

wentam committed:
pine64-pinephonepro: kernel support for keyboard

wentam committed:
pine64-pinephonepro: dts fix for wireplumber

wentam committed:
pine64-pinephonepro: kernel patch for wake on call

samueldr commented:
From https://github.com/NixOS/mobile-nixos/pull/522#discussion_r1007811259

> > the default setup works with "stage-0", which in turn means the boot image is not required to be updated, since the system will `kexec` into the generation's kernel.
> 
> This had me somewhat confused at first as my system was clearly not doing this. Looks like this is just because the pinephone pro isn't currently set up this way, and setting `mobile.quirks.supportsStage-0 = true` results in a kernel panic on boot.
> 
> Makes sense. Sounds like I should be figuring out how to get stage-0 style boot working with pinephone pro.

Right, this might be an issue, as `kexec` support was reportedly problematic on RK3399 some time back, and I don't know what is the expected status on the Pinephone Pro.

wentam commented:
> Right, this might be an issue, as `kexec` support was reportedly problematic on RK3399 some time back, and I don't know what is the expected status on the Pinephone Pro.

Played with this for a bit. Booting the first kernel with `irqchip.gicv3_nolpi=1` (not needed for generation kernel) gets past the kernel panic ([reference](https://patchwork.kernel.org/project/linux-arm-kernel/cover/20180921195954.21574-1-marc.zyngier@arm.com/)). Unfortunately generation kernel just hangs after this and I haven't found a good way to get a stack trace at this point.

I'm not very familiar with kernel development and will likely be letting this be for now.

wentam committed:
pine64-pinephonepro: kernel 5.17.9 -> 6.0.7

wentam commented:
- kernel 5.17.9->6.0.7
- waydroid and firewall commits removed in favour of #536 and #542
- stage-0 style boot now working

wentam committed:
pine64-pinephonepro: Add README

wentam committed:
pine64-pinephonepro: kernel 6.0.7 -> 6.0.8

wentam commented:
* kernel 6.0.7->6.0.8
* Added a README based on pinephone's

zhaofengli committed:
pine64-pinephonepro/kernel: Enable uinput

zhaofengli committed:
pine64-pinephonepro: Add firmware to system profile

zhaofengli committed:
pine64-pinephonepro/kernel: Enable loading compressed firmware

zhaofengli commented:
I dusted off my PinePhone Pro and tried this. I installed the phone fresh with the installer image, and with stage-0 enabled and the 6.0.8 kernel in the boot partition, after kexec'ing the generation 6.0.8 kernel almost always gets stuck:

```
[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 6.0.8 (nixbld@localhost) (gcc (GCC) 9.5.0, GNU ld (GNU Binutils) 2.39) #1-mobile-nixos SMP PREEMPT Tue Jan 1 00:00:00 UTC 1980
[    0.000000] Machine model: Pine64 PinePhonePro
[    0.000000] earlycon: uart8250 at MMIO32 0x00000000ff1a0000 (options '')
[    0.000000] printk: bootconsole [uart8250] enabled
[    0.007550] interrupt-controller@fee20000: unable to locate ITS domain
[   22.877287] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[   22.884120] rcu:     2-...!: (0 ticks this GP) idle=67c4/0/0x1 softirq=204/204 fqs=0
[   22.892518] rcu:     3-...!: (9 GPs behind) idle=46b4/1/0x4000000000000000 softirq=26/26 fqs=0
[   22.901881] rcu:     4-...!: (0 ticks this GP) idle=24a0/0/0x0 softirq=41/41 fqs=0 (false positive?)
[   22.912159] rcu: rcu_preempt kthread timer wakeup didn't happen for 21002 jiffies! g-423 f0x0 RCU_GP_WAIT_FQS(5) ->state=0x402
[   22.924915] rcu:     Possible timer handling issue on cpu=2 timer-softirq=193
[   22.932610] rcu: rcu_preempt kthread starved for 21005 jiffies! g-423 f0x0 RCU_GP_WAIT_FQS(5) ->state=0x402 ->cpu=2
[   22.944296] rcu:     Unless rcu_preempt kthread gets sufficient CPU time, OOM is now expected behavior.
[   22.954513] rcu: RCU grace-period kthread stack dump:
[   22.960395] rcu: Stack dump where RCU GP kthread last ran:
```

So I'm currently using this with stage-0 disabled in my setup. Furthermore, to get Wi-Fi upon resume to work I had to [enable `CONFIG_FW_LOADER_COMPRESS`](https://github.com/zhaofengli/mobile-nixos/commit/274541310152e33104c5e45dc1c4b2a509972b2e) like in #508 and [add the firmware to system profile](https://github.com/zhaofengli/mobile-nixos/commit/f03913590376fd5bede6583dd47edbd4e14bdf5c).

Here is my [entire diff](https://github.com/zhaofengli/mobile-nixos/compare/d1b8b0dd00ffbcd559be9f0ddce704250bd9620b...f03913590376fd5bede6583dd47edbd4e14bdf5c).

samueldr commented:
I'm thinking we probably should enable adding `config.mobile.device.firmware` to `hardware.firmware` automatically by default, as it seems to be a common pain point among users.

The only reason I don't notice it myself is because the few devices I use with Mobile NixOS I use my "integrated" custom config where I do this outright.

(To be handled in a separate PR)

wentam commented:
Added @zhaofengli's commits (thanks!)

6.0.7->6.0.8 and 6.0.8->6.0.8 kexec is working reliably here. I'm curious to know if anyone else has @zhaofengli's crash (and with what boot partition and generation kernel versions). If kexec boot isn't going to be reliable, we'd be better off setting this up to have the kernel manually updated as before (until it's time to play with EFI boot).

tejing1 commented:
Just installed from this, and kexec boot hasn't cause me an issue thus far, other than that I have to enter my disk password twice on boot, of course (though I've only rebooted 3-4 times).

Issues I do have:
- A flickering/bouncing effect in phosh. Not sure if this is purely a phosh bug or if it's something in the kernel video driver. Things near the bottom of the screen get drawn lower than their actual position inconsistently from one frame to the next, and black bars appear above them, like maybe an output buffer is underrunning somewhere. It's only present sometimes, though.
- Probably not related to this PR, but the on screen keyboard for entering the disk password is very inconsistent about recognizing touch events. I really have to baby it and watch every character to make sure it actually takes. I know this isn't an actual hardware issue since the on screen keyboard in phosh is much better.

iclanzan commented:
I have also successfully built, flashed and booted the PPP using this code and am also experiencing all of the issues @tejing1 mentioned.

The phosh glitches are not happening for me when not kexec’ing so I am wondering if it might be related. Or maybe the different kernel version is to blame.

Is there no way to prevent being asked for the luks password twice when kexec’ing?

tejing1 commented:
Tested without kexec. I also see no phosh glitches in that case.

kchibisov commented:
>Tested without kexec. I also see no phosh glitches in that case.

The glitches are not related to phosh, since I have them in any other compositor(sway, catacomb). And yes, disabling kexec fixes them.

samueldr commented on devices/pine64-pinephonepro/README.adoc:
```diff
@@ -0,0 +1,90 @@
+
+=== Platform Firmware
+
+It is recommended to link:https://tow-boot.org/devices/pine64-pinephonePro.html[install Tow-Boot to SPI].
```
```suggestion
The only *supported* platform firmware configuration is to link:https://tow-boot.org/devices/pine64-pinephonePro.html[install Tow-Boot to SPI].
```

Let's make it obvious it is not only recommended, but the only supported method.

samueldr commented on devices/pine64-pinephonepro/default.nix:
```diff
@@ -0,0 +1,123 @@
+    "earlycon=uart8250,mmio32,0xff1a0000"
+    "earlyprintk"
+
+    "quiet"
+    "vt.global_cursor_default=0"
+
```
```suggestion
```

https://github.com/NixOS/mobile-nixos/pull/525 obsoletes those.

samueldr commented on devices/pine64-pinephonepro/default.nix:
```diff
@@ -0,0 +1,123 @@
+  };
+
+  boot.kernelParams = [
+    # Serial console on ttyS2, using the dedicated cable.
+    "console=ttyS2,115200"
+    "earlycon=uart8250,mmio32,0xff1a0000"
+    "earlyprintk"
+
```
```suggestion
    # Serial console on ttyS2, using the dedicated cable.
    # TODO: option to enable serial console.
    #"console=ttyS2,115200"
    #"earlycon=uart8250,mmio32,0xff1a0000"
    #"earlyprintk"

```

For now comment these out like done e.g. here:

 - https://github.com/NixOS/mobile-nixos/blob/4d2093efa7efa00131d385fd9d11e54ce16bc57a/devices/lenovo-wormdingler/default.nix#L41-L44

Add the `TODO` comment as well. Eventually (haha) someone (me?) will come back around and add an option to describe the required config for the serial console, which will be controlled behind an `.enable` option.

samueldr commented on devices/pine64-pinephonepro/default.nix:
```diff
@@ -0,0 +1,123 @@
+      INPUT_GPIO_VIBRA = yes;
+    })
+  ];
+
+  mobile.boot.stage-1.shell.console = "ttyS2";
+  mobile.boot.stage-1.shell.shellOnFail = true;
```
```suggestion
```

Though I guess I had those enabled in my PR (haven't checked), these are for debugging, and should not be left in the config. (Though relatedly to the other console stuff, the stage-1 shell console should be configured in that other option set I suppose.)

samueldr commented on devices/pine64-pinephonepro/kernel/default.nix:
```diff
@@ -0,0 +1,34 @@
+  };
+
+  patches = [
+    ./0001-HACK-always-show-logo.patch
```
```suggestion
    ./0001-HACK-always-show-logo.patch
```

(And remove the file)

I know I added this patch, I'm not sure we should enable it outright in the end. (After all this was a really-WIP branch I published at the time.)

Or if it is desirable, I think this shouldn't be part of a specific device since this is an overall problem, not a Pinephone Pro problem.

samueldr commented on modules/hardware-rockchip.nix:
```diff
@@ -12,6 +11,11 @@ in
+    hardware.socs.rockchip-rk3399s.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = "enable when SOC is RK3399";
```
```suggestion
      description = "enable when SOC is RK3399S";
```

woopsie

samueldr reviewed (commented):
I've been reading these changes. Most are from decisions from my past self. Though some are from how I've began steering the project, so no one knew (not even me) they would be required at some point :).

samueldr commented:
Other than that, we'll want to add the installer to the release files, including `mobile.documentation.hydraOutputs` which I now realize I forgot on the other devices Ia dded support for recently.

I think otherwise it's working as well as it would work elsewhere? Any issues with the phone environments (DEs) is unrelated to this PR, I suspect, and otherwise anything else is blocked on upstream, right?

samueldr commented:
(Re-reading your initial description) AFAIK stk3311 failure comes from a Platform Firmware issue, there is an issue (and even PR I believe) in Tow-Boot, I need to take time and circle back around all Tow-Boot things too.

tejing1 commented:
>Any issues with the phone environments (DEs) is unrelated to this PR, I suspect

The graphical issues are tied to the use of kexec, so it's worth considering if we should disable stage0 support.

iclanzan commented:
The modem is not working in this PR, it needs a service like the one [for PinePhone)(https://github.com/NixOS/mobile-nixos/blob/master/devices/pine64-pinephone/modem.nix) to power it on, right?

samueldr commented:
> > Any issues with the phone environments (DEs) is unrelated to this PR, I suspect
> 
> The graphical issues are tied to the use of kexec, so it's worth considering if we should disable stage0 support.

I meant other than that, but yeah, I wasn't clear. Though I didn't see any issues on Plasma Mobile. So reality may lie somewhere in-between.

samueldr commented:
> The modem is not working in this PR, it needs a service like the one [for PinePhone](https://github.com/NixOS/mobile-nixos/blob/master/devices/pine64-pinephone/modem.nix) to power it on, right?

Right, I should have stated that I wasn't set-up to test the modem, and didn't. If it's trivial, we can get it going quickly here, or else have it handled within another PR.

I think there is still the overall issue of reworking the Pinephone (A64)'s modem handling, which I support the same implementation can be used on the Pinephone Pro? I'm totally out of touch with these parts at the moment.

wentam committed:
pine64-pinephonepro: kernel 6.0.8 -> 6.1.10

wentam committed:
pine64-pinephonepro: disable stage-0/kexec boot

kexec still creates issues in latest kernel.

wentam committed:
pine64-pinephonepro: serial console off by default

wentam committed:
pine64-pinephonepro: Remove obsolete kernel params

wentam committed:
pine64-pinephonepro: Remove stage-1 debug options

wentam commented:
> > The modem is not working in this PR, it needs a service like the one [for PinePhone](https://github.com/NixOS/mobile-nixos/blob/master/devices/pine64-pinephone/modem.nix) to power it on, right?
> 
> Right, I should have stated that I wasn't set-up to test the modem, and didn't. If it's trivial, we can get it going quickly here, or else have it handled within another PR.
> 
> I think there is still the overall issue of reworking the Pinephone (A64)'s modem handling, which I support the same implementation can be used on the Pinephone Pro? I'm totally out of touch with these parts at the moment.

I've been running eg25-manager in my local configuration as the eg25-manager efforts exist elsewhere than this PR. Once #573 or similar is merged, adding it here should be a one-liner. No reason for that to be a blocker for this PR IMO.

> 
> Though it is ticked in the original PR description, so I'm not sure what to make of that.

This was ticked to represent that the hardware was generally working, not the userspace. I was able to make/receive phone calls etc (with some userspace-related glitches) on my system with eg25-manager running.

wentam committed:
release.nix: Add pine64-pinephonepro installer

wentam committed:
pine64-pinephonepro: Add mobile.documentation.hydraOutputs

wentam committed:
pine64-pinephonepro: Update README to follow kexec boot disable

wentam commented:
* kernel 6.0.8 -> 6.1.10
* Made @samueldr's requested changes
* Disabled kexec/stage-0 boot - it's probably possible to make it play along, but it's going to continue to be high-maintenance as different issues appear with different kernel versions. Considering EFI plans, it's not worth spending the time to deal with the fragility IMO.
* Updated README to reflect that you now need to manually update the boot partition/kernel
* Added to release.nix and added mobile.documentation.hydraOutputs

There will always be more to play with:
* No idea if the cameras are working, but I think they should in theory in current kernel version. Current userspace still doesn't support them, so this is mildly annoying to find out. Given the lack of userspace support, there's no reason for this to be a blocker.
* STK3311 still doesn't show up. This will probably improve with the next release of Tow-Boot.
* eg25-manager needs to get added for modem usage, this will be very simple once eg25-manager is available in the mobile-nixos repo.

Overall, though, I think this PR is in a pretty sane place right now - and it will be easier for others to contribute once this is merged.

samueldr commented:
(Adding a quick comment before even looking at it again)

I agree that we shouldn't block for that last list of issues, they're not issues that are caused by the integration with Mobile NixOS, they're upstream/downstream issues (depending).

tejing1 commented:
I just rebased off of this, and it doesn't actually build:
```
error: The ‘env’ attribute set cannot contain any attributes passed to derivation. The following attributes are overlapping: NIX_CFLAGS_COMPILE
```
It seems to be coming from `boot/script-loader/mruby-lvgui-native/lvgui.nix`, if I'm reading the trace correctly.

At any rate, I'm certain the failure is unrelated to my own commits.

wentam commented:
> I just rebased off of this, and it doesn't actually build (against nixpkgs commit 68196a61c26748d3e53a6803de3d2f8c69f27831):
> 
> ```
> error: The ‘env’ attribute set cannot contain any attributes passed to derivation. The following attributes are overlapping: NIX_CFLAGS_COMPILE
> ```
> 
> It seems to be coming from `boot/script-loader/mruby-lvgui-native/lvgui.nix`, if I'm reading the trace correctly.
> 
> At any rate, I'm certain the failure is unrelated to my own commits.

Perhaps related to #580?

tejing1 commented:
That certainly seems to be the same issue. Sounds like it's non-trivial to fix properly.

Well, for now, I'll just stick with an older nixpkgs and worry about it later.

samueldr commented:
Fix now in master... Luckily [I could just ignore the problem entirely](https://github.com/NixOS/mobile-nixos/pull/585).

stites commented:
Seems to be working for me! Thank you!

if i want camera support, should I rebuild the iso on my current ppp, then reflash with the natively compiled iso?

cameronfyfe commented:
This is working for me today, thanks for the work you've done!

I am a bit confused about how the modem works though. I saw that I had no network connection through the modem and then saw in this thread that that was not part of this PR, so I grabbed the eg25-manager package from #573, ran that and then was able to connect and send some SMS messages. However now I've rebooted and I still have a working 4g connection without eg25-manager running.

Does anyone know how this is setup? Is eg25-manager only required for some initialization, but the main AT interface to the modem is through a driver already included?

aszlig commented on devices/pine64-pinephonepro/kernel/config.aarch64:
```diff
@@ -0,0 +1,6669 @@
+# CONFIG_XFS_FS is not set
+# CONFIG_GFS2_FS is not set
+# CONFIG_OCFS2_FS is not set
+# CONFIG_BTRFS_FS is not set
```
What's the reason for disabling this? I'm mainly asking since I found that the config for the normal PinePhone has this [enabled](https://github.com/NixOS/mobile-nixos/blob/ba4638836e94a8f16d1d1f9e8c0530b86078029c/devices/pine64-pinephone/kernel/config.aarch64#L7071) and also because I just had to patch this for installing with BTRFS.

samueldr replied:
There is no rationale, other than the config files being started from "well known" defconfig files...

They could and can be enabled. Though there is an overall problem that still needs solving, and it is dealing with the generation's own kernel; once that would be handled, the kernel semantics would change and we'd mostly use the NixOS config (sprinkled with patches and additional configs to enable as needed).

samueldr replied:
I'll add a quick note that since #651, a fresh change, we are now enabling more filesystems by default.

 - https://github.com/NixOS/mobile-nixos/blob/development/support/kernel-config/filesystems.nix

It's a simpler solution than my unattainable (for now) ideal. Feel free to suggest others that may be generally needed and missing!

Though note that now it is possible to use the `structuredConfig` attrset for personalization, so specialized use-case don't need to change defaults anymore, and yet stay ergonomic enough.

bendlas commented:
I've been running a configuration based on this on my RockPro64 for a couple of months. Just rebased on @aszlig's branch, and with kernel 6.3: https://github.com/bendlas/mobile-nixos/blob/rockpro64/devices/pine64-rockpro64/default.nix

The pinephone pro and rockpro64 are very similar. I mainly just needed to add drivers for Ethernet and for HDMI Audio.

Is there a proper way in mobile-nixos to share hardware platforms, or should I just go back to copy&paste everything, as [before my latest rebase](https://github.com/bendlas/mobile-nixos/blob/ppp/devices/pine64-rockpro64/default.nix)?

wentam committed:
pine64-pinephonepro: kernel 6.1.10 -> 6.4.7

wentam commented:
* rebase
* kernel 6.1.10 -> 6.4.7

STK3311 is working now with latest tow-boot ([2022.07-006](https://github.com/Tow-Boot/Tow-Boot/releases/tag/release-2022.07-006))

Gooberpatrol66 commented:
I built [d08999b](https://github.com/NixOS/mobile-nixos/commit/d08999b9317a1de0fd732545411e754409071444) for plasma mobile and booted it on my ppp. The wifi, bluetooth, and flashlight don't work.

wentam commented:
> I built [d08999b](https://github.com/NixOS/mobile-nixos/commit/d08999b9317a1de0fd732545411e754409071444) for plasma mobile and booted it on my ppp. The wifi, bluetooth, and flashlight don't work.

Wifi, bluetooth, flashlight all working for me at the same revision and phosh. Is it possible it's a plasma mobile issue or something else perhaps?

Gooberpatrol66 commented:
> > I built [d08999b](https://github.com/NixOS/mobile-nixos/commit/d08999b9317a1de0fd732545411e754409071444) for plasma mobile and booted it on my ppp. The wifi, bluetooth, and flashlight don't work.
> 
> Wifi, bluetooth, flashlight all working for me at the same revision and phosh. Is it possible it's a plasma mobile issue or something else perhaps? Does the previous revision ([13e69a8](https://github.com/NixOS/mobile-nixos/commit/13e69a87a85cdc238dba916d6dc5280502f7d9c8)) exhibit the same issues?

ip link doesn't show the wlan device, only usb and lo. it seems like a lower level issue

wentam commented:
> 



> > > I built [d08999b](https://github.com/NixOS/mobile-nixos/commit/d08999b9317a1de0fd732545411e754409071444) for plasma mobile and booted it on my ppp. The wifi, bluetooth, and flashlight don't work.
> > 
> > 
> > Wifi, bluetooth, flashlight all working for me at the same revision and phosh. Is it possible it's a plasma mobile issue or something else perhaps? Does the previous revision ([13e69a8](https://github.com/NixOS/mobile-nixos/commit/13e69a87a85cdc238dba916d6dc5280502f7d9c8)) exhibit the same issues?
> 
> ip link doesn't show the wlan device, only usb and lo. it seems like a lower level issue

Anything interesting in dmesg (`dmesg | grep -i wlan`)? Worth checking the dip switches just in case that got bumped, too.

Further debugging might be easier in the matrix channel (see README for this repo).

Gooberpatrol66 commented:
[dmesg.txt](https://github.com/NixOS/mobile-nixos/files/12300699/dmesg.txt)

Gooberpatrol66 commented:
Are there binary blobs that I need to get from somewhere to put in the build?

wentam commented:
> [dmesg.txt](https://github.com/NixOS/mobile-nixos/files/12300699/dmesg.txt)
> 
> The switches are fine, it all works in mobian
> 
> There's some text that flashes during the nixos mobile logo, not sure what that is something about "firmware attach blob not found"
>
> Are there binary blobs that I need to get from somewhere to put in the build?

Firmware blobs are in this PR and you shouldn't need to do anything special (https://github.com/NixOS/mobile-nixos/blob/d08999b9317a1de0fd732545411e754409071444/devices/pine64-pinephonepro/firmware/default.nix).

I do get some firmware loading errors in dmesg as well on my working system, so that could be a red herring.

Thought: 6.4 kernel did change some rk8xx stuff, it stands to reason that would be responsible for controlling power to the radios (though I didn't actually check so don't quote me on that).

If you're able to spend the time, building 13e69a87a85cdc238dba916d6dc5280502f7d9c8 would isolate the kernel update and let us know if the issue lives there.

Gooberpatrol66 commented:
> > [dmesg.txt](https://github.com/NixOS/mobile-nixos/files/12300699/dmesg.txt)
> > The switches are fine, it all works in mobian
> > There's some text that flashes during the nixos mobile logo, not sure what that is something about "firmware attach blob not found"
> > Are there binary blobs that I need to get from somewhere to put in the build?
> 
> Firmware blobs are in this PR and you shouldn't need to do anything special (https://github.com/NixOS/mobile-nixos/blob/d08999b9317a1de0fd732545411e754409071444/devices/pine64-pinephonepro/firmware/default.nix).
> 
> I do get some firmware loading errors in dmesg as well on my working system, so that could be a red herring.
> 
> Thought: 6.4 kernel did change some rk8xx stuff, it stands to reason that would be responsible for controlling power to the radios (though I didn't actually check so don't quote me on that).
> 
> If you're able to spend the time, building [13e69a8](https://github.com/NixOS/mobile-nixos/commit/13e69a87a85cdc238dba916d6dc5280502f7d9c8) would isolate the kernel update and let us know if the issue lives there.

I also built older versions like [e3a600c](https://github.com/aszlig/mobile-nixos/commit/e3a600c8378e76acbf5866ca752a6776c7ae9709) and it was still broken.

I built a phosh version and the flashlight and bluetooth works, but the wifi is still broken
[dmesg.txt](https://github.com/NixOS/mobile-nixos/files/12307251/dmesg.txt)

samueldr committed:
pine64-pinephonepro: Normalize kernel config

(Automatically using bin/kernel-normalize-all)

samueldr committed:
pine64-pinephonepro: Make use of mobile.boot.serialConsole

samueldr committed:
pine64-pinephonepro: Re-order contents

Follow the Pinephone (A64) order, for now.

We probably should somehow have a stricter order.

samueldr committed:
pine64-pinephonepro: Move structured kernel config requirements to a file

We probably want to do this, per-family first, then per-device. It's a
neat way to ensure that requirements are kept, and renamed options don't
cause issues down the line.

samueldr committed:
pine64-pinephonepro: Mark as supported

samueldr commented:
Other than my inaction (sorry!), and a rebase (which I've done already)...

Is there anything technical still preventing this to be merged?

The rebase + misc. updates is here:

 - https://github.com/samueldr-wip/mobile-nixos-wip/tree/PR535

It makes use of the new feature for configuring the serial console, and otherwise, marks it as a supported device. Other than that it's morally the same set of changes, but rebased on top of current `development`.

Minor issues we should aim to fix in subsequent PRs, as it's obvious since... last year... that it will never be merged otherwise.

wentam commented:
> Is there anything technical still preventing this to be merged?

So long as everything appears to you to be structurally correct/following repo conventions, I think this is in a good place to be merged. To me, everything supported in the kernel we're using appears to work. There are probably some minor issues at this stage, but getting something structurally correct merged will only serve to make it easier to manage any issues on an individual basis (such as enabling eg25-manager via #573).
